### PR TITLE
Create structs to represent C++ classes & vtables by allowing unsized structs

### PR DESCRIPTION
**Pull request by zbanks**
This is a follow-up from a Discord conversation from a few days ago. 
This isn't a particularly small change, so I'm on the fence if this should get merged right now.

- Create structs for vtables by parsing the symbol names in the asm data.
    - Awkwardly done in translate.py for now, because it needs access to asm data recursively. (We also want to avoid re-parsing the types for the symbols referenced by the vtable.)
- To really benefit from this, it requires creating structs to represent C++ classes so the vtable field names can be used in other functions. However,we don't know the size of the classes.
- This is awkward to represent with `Type`/`TypeData`, because the size of these structs is not kept consistent with `TypeData.size_bits`.

Some example output from Pikmin 2: [gist](https://gist.github.com/zbanks/86c3774f34dac8eea0a89c5531cb82d4) (I don't think a diff is useful here)

### CODE REVIEW

File: src/types.py [hunk 1]
```diff
@@ -995,10 +1029,12 @@ def demangled_type(sym_type: CxxType, sym_name: Optional[CxxTerm] = None) -> "Ty
                     and len(sym_name.qualified_name) > 1
                     and final_name not in CxxSymbol.STATIC_FUNCTIONS
                 ):
+                    class_type = type_for_class(sym_name.qualified_name[:-1])
                     # NB: This assumes `this` is passed as the first arg, which is true
                     # for most thiscall methods on PPC. However, this is incorrect for
                     # static member functions, functions returning structs, and other ABIs.
-                    params.append(FunctionParam(type=Type.ptr(), name="this"))
+                    params.append(FunctionParam(type=Type.ptr(class_type), name="this"))
+                    # TODO: Virtual methods may take a second argument here
```
Comment (simonlindholm) [inline src/types.py:1037]: what second argument? this comment is going to confuse me in the future

File: src/types.py [hunk 2]
```diff
@@ -309,7 +320,7 @@ def get_parameter_size_align_bytes(self) -> Tuple[int, int]:
         data = self.data()
         if self.is_struct():
             assert data.struct is not None
-            return data.struct.size, data.struct.align
+            return data.struct.min_size(), data.struct.align
```
Comment (simonlindholm) [inline src/types.py:323]: This is rather sketchy. Should probably throw a DecompFailure(f"Struct parameter of incomplete type {data.struct}") instead, not that it's likely to ever come up in practice

**Review by simonlindholm — COMMENTED** [review]
vtable accesses are done with x->unk_0->foo() / (*x)->foo(), would be nice if we could cut out the unk_0-> part or at least rename unk_0 to indicate a vtable. But that can be future work.

Definitely seems worth merging this to me.

**Comment by simonlindholm** [comment]
also, could be worth leaving the vtable data out of the output since you're not going to define that manually in C anyway. Or putting it behind some "verbose" cmdline flag

File: src/types.py (hunk 1 above)
Comment (zbanks) [reply src/types.py]: I'm now confident that this was a mistake to begin with. The issue has nothing to do with virtual methods, I think it is just that when the function returns a struct/class, then there's an argument with a pointer to that return value (and that argument nor the return value is indicated in the mangled name).

File: src/translate.py [hunk 3]
```diff
@@ -4544,6 +4544,51 @@ def address_of_gsym(self, sym_name: str) -> AddressOf:
...
+                for entry in sym.asm_data_entry.data:
+                    if isinstance(entry, bytes):
+                        if len(entry) % 4 != 0:
+                            raise DecompFailure(
+                                "Unable to parse misaligned vtable data in {sym_name}"
+                            )
+                        for i in range(len(entry) // 4):
+                            field_name = f"{vt_struct.new_field_prefix}{offset:X}"
+                            vt_struct.try_add_field(
+                                Type.ptr(), offset, field_name, size=4
```
[the remark this answers is not in the document]
Comment (zbanks) [reply src/translate.py]: I've only found `0`'s so far, but [this blog post](http://hacksoflife.blogspot.com/2007/02/c-objects-part-3-multiple-inheritance.html) indicates that the vtable should start with a typeid pointer, followed by an offset. (I guess if the typeinfo stuff is disabled, it makes sense that I'd only see `NULL, 0` here)

File: src/translate.py [hunk 4]
```diff
@@ -4544,6 +4544,51 @@ def address_of_gsym(self, sym_name: str) -> AddressOf:
...
+                vt_size = sym.asm_data_entry.size_range_bytes()[1]
+                vt_struct = StructDeclaration.unknown(
+                    self.typepool, size=vt_size, align=4, tag_name=sym_name
+                )
+                offset = 0
+                for entry in sym.asm_data_entry.data:
+                    if isinstance(entry, bytes):
+                        if len(entry) % 4 != 0:
+                            raise DecompFailure(
+                                "Unable to parse misaligned vtable data in {sym_name}"
```
[the remark this answers is not in the document]
Comment (zbanks) [reply src/translate.py]: Yeah, I make this mistake *embarrassingly* often :facepalm:

**Comment by zbanks** [comment]
> vtable accesses are done with x->unk_0->foo() / (*x)->foo(), would be nice if we could cut out the unk_0-> part or at least rename unk_0 to indicate a vtable. But that can be future work.

Addressed the inline feedback, but I don't have any particularly good idea for how to do this part.

File: src/types.py [hunk 5]
```diff
@@ -1330,21 +1373,26 @@ def pad_to(offset: int, is_final: bool) -> None:
                 if not field.known:
                     comments.append("inferred")
                 lines.append(fmt.with_comments(field_decl, comments))
+                # Hack to recompute field.type's size
```
[the remark this answers is not in the document]
Comment (zbanks) [reply src/types.py:1393]: (I'm going to punt this for later)

**Review by simonlindholm — APPROVED** [review]

**Pull request merged**